platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,341 @@
+By doing so both metrics can use the same name but are still unique as long as the labels are different.
+We will store measurements individually for each metric.
+If you later want to know the combined measurement of all metrics with the name `api_http_requests_total` you can use the prometheus query language to sum up the values of all metrics with that name.
+See https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors for more information.
+
+In prometheus a concrete metric is defined like this (see https://prometheus.io/docs/concepts/data_model/#notation):
+
+```
+api_http_requests_total{method="POST", handler="/messages"}
+```
+
+Here a metric is defined by a name (`api_http_requests_total`) and a set of labels (`method="POST"` and `handler="/messages"`).
+In prometheus that defines the unique identifier of a metric.
+
+When using labels to create dashboards in Grafana prometheus provides a rich query language to filter and group metrics: https://prometheus.io/docs/prometheus/latest/querying/examples/
+
+We will use the same practices and definition as in prometheus regarding labels.
+
+### Public API
+
+The metrics module will be extended by a new data type called `Label`.
+A label is a key-value pair that can be defined as a `record`:
+
+```java
+record Label(@NonNull String key, @Nullable String value) {}
+``` 
+
+> [!NOTE]  
+The value of a label can be `null`.
+For our prometheus endpoint a label with a `null` value will be handled as if the label would not exist.
+This would allow us to define a label without a value and with a possible extension (see "Future Work") we define a concrete value for a label later.
+Today best practice is to not use labels with a `null` value.
+
+The `Metric` interface will be extended by a new method called `getLabels()` that returns a list of labels:
+
+```java
+interface Metric {
+    
+    //...
+    
+    @NonNull
+    Set<Label> getLabels();
+}
+```
+
+Like the name or category, the labels are defined when a metric is created and cannot be changed later.
+Based on that the MetricConfig class will be extended by a new method called `withLabels`:
+
+```java
+abstract class MetricConfig {
+    
+    //...
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Label... labels) {...}
+    
+    @NonNull
+    MetricConfig withLabels(@NonNull Set<Label> labels) {...}
+}
+```
+
+By adding labels the unique identifier of a metric is the name and the labels.
+Based on that we need to modify the `Metrics` interface to use the labels as the unique identifier of a metric:
+
+```java
+
+interface Metrics {
+    
+    // old methods:
+    // Object getValue(@NonNull String category, @NonNull String name);
+    // void remove(@NonNull String category, @NonNull String name);
+    
+    @Nullable
+    Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
```

Comment:
If we are going to include labels as part of the unique id, then it makes sense to me to have a`MetricId` record with these values in it. Then it is very clear what is part of a metric's unique identifier.